nixos/bird: rename bird2 to bird, switch to bird3 by default

herbetom opened the pull request:
Now that there is a version 3 of bird i think we should cleanup some bird2 stuff and change the default bird package to v3.

In light of that i'm proposing to do a couple of things:

- renaming the current `bird` package to `bird2`
- introducing a `bird3` package
- having an `bird` package alias which points to `bird3`
- renaming the `services.bird2` options to `services.bird`
- introducing a `services.bird.package` option to allow changing the new bird3 default

I would still keep bird2 around because as of now it seems to have a lower memory footprint and users could be required to adjust their config. And with both versions beeing supported officially (as of now, no word yet how long this will last), we might as well have both for the time beeing to enable a smooth transition.


The bird project has some notes on the migration to v3 here:
https://gitlab.nic.cz/labs/bird/-/blob/v3.0.1/doc/migration-bird3.md

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

herbetom committed:
nixos/bird: rename bird2 to bird, switch to bird3 by default

This is done in view of the Release of the new v3 of Bird.

Switch to the bird3 package for the `services.bird.package` option.

Switch the `bird` package alias to bird3.

herbetom committed:
bird: rename bird to bird2

This is done to enable us to introduce bird3 without causing confusion.

herbetom committed:
bird3: init at 3.0.1

herbetom commented:
Not sure wheter this should be classified as a `significant` or a `major or breaking` change. I would appreciate input in that regard.

herbetom commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366190`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bird2</li>
    <li>bird3</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bird2</li>
    <li>bird3</li>
  </ul>
</details>

herbetom commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366190`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bird2</li>
    <li>bird3</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bird2</li>
    <li>bird3</li>
  </ul>
</details>

snaakey commented:
Defaulting to bird3 will break the configuration of some people. I would suggest to default to bird2 for now and throw a warning when the package option is not explicitly set and `checkConfig` is disabled. So that people get notified that the default package will change in the future and break their configuration if they don't explicitly specify the package.

herbetom committed:
nixos/bird: add package option

This is done to allows to easier change which bird package should be used

herbetom commented:
@snaakey Thanks for taking a look! Having a check requiring Users to set the package option when `checkConfig` is set to false sounds like an intersting idea. I'll look into it.

My worry with keeping bird2 as the default is that it's unclear how long both versions will be maintained. Probably long enough for the rather short stable window of NixOS, but still, i would prefer to do the switch for 25.05 and be done with it.
And with us switching from `services.bird2` to `services.bird` it already requires manual intervention by every single user and that should already be quite a warning that something is happening. Switching later doesn't have that same benefit and Users with checkConfig enabled would be less informed that the switch is happening.

herbetom commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366190`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.bird</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bird2</li>
    <li>bird3</li>
  </ul>
</details>

SuperSandro2000 commented on nixos/modules/services/networking/bird.nix:
```diff
@@ -68,36 +69,38 @@ in
   imports = [
-    (lib.mkRemovedOptionModule [ "services" "bird" ] "Use services.bird2 instead")
-    (lib.mkRemovedOptionModule [ "services" "bird6" ] "Use services.bird2 instead")
+    (lib.mkRemovedOptionModule [ "services" "bird2" ]
```
Why not use mkRenamedOptionModule and make package depending on stateVersion?

fooker replied:
Using `mkRenamedOptionModule` is arguable but it can not give a custom user message.
Depending on `stateVersion` is strange as the user has to handle the change anyways and the added `package` option seems to be enough.

CutestNekoAqua replied:
because its not renamed but actually changed in this case. and having a general services.bird makes more sense in the first place as thats the general practice on nixpkgs in general, this is just bikeshedding imo and preventing useful things from happening on nixpkgs.

SuperSandro2000 replied:
> Depending on `stateVersion` is strange as the user has to handle the change anyways and the added `package` option seems to be enough.

If we are using mkRenamedOptionModule and depend on stateVersion then a user doesn't have to do anything immediately and things just continue to work.

If we do it like the current proposal, any user using any of bird has to rename everything and set package to just have their setup continue to work OR migrate to bird3. Both variants are arguable bad in the sight of a "I'll fix the warning later" option.

> because its not renamed but actually changed in this case.

No, you're wrong. If we depend on stateVersion like in many other modules eg: nextcloud or postgresql then there is no change.

> having a general services.bird makes more sense in the first place

Yep but that is not my point.

herbetom replied:
> No, you're wrong. If we depend on stateVersion like in many other modules eg: nextcloud or postgresql then there is no change.

If we were to just pin the package there still would be changes:
- The name of the systemd service
- The user/group as which it's executed
- the location of the config file

For most Users this probably isn't relevant, but changing those still could break something somewhere.
One could now attempt to pin all of those values to the stateVersion but that appears to me like enourmus complexicty and a testing and validation nightmare going forward.

herbetom replied:
Depending on the stateVersion also has the drawback that in a fleet of old as well as new systems i would find it unexpected that with the same config (and stable release) one system runs one major version and another the other. 
And with bird config i would find it highly likely that users will share config between multiple systems but since config in most cases probably validates with both versions issues may be quite subtle.

snaakey replied:
> Why not use mkRenamedOptionModule and make package depending on stateVersion?

Letting the Package depend on `stateVersion` is a horrible idea since all changes are completely config related and `mkRenamedOptionModule` won't tell users that changing this will implicitly also update to the newest major version of bird. In my opinion it should only be used for stuff that effects _system state_ that cannot be controlled by the config directly. Letting stuff depend on `stateVersion` is mostly implicit and silent which is just a nightmare. Really `stateVersion` should only be used if there is no other option in my imo.

MarcelCoding replied:
Alternatively, one could just omit the package option and force the user to explicit chose the package and thereby the version. The MySQL module is doing that to force the choice between MySQL and MariaDB.

herbetom replied:
@MarcelCoding I think that would be more appropiate if we would expect there to be a "bird4" rather shortly. But if that ever happens i would expect it to be quite a few years. After all, bird3 was released 7 years (and 6 days) after bird 2.0.0 and another somewhat breaking release (bird 2 to 3 doesn't even neccearily really have to cause any issues) is probably also going to be quite some time away.
Further i would expect the EOL of bird2 to happen quicker then it was the case with bird1 (~6 years after 2.0.0) as there shouldn't be that much of a difference and upgrade hassle.

My current best guess is therefore that there will be at least a couple of years where only bird3 will be supported. (But of course who knows ...).

Accordingly i think we should just set bird3 as the default and if there is ever the need to bump that we can still nag people to set something or just bump it and only complain if checkConfig isn't true or whatever. But that is something i would decide only once we know what the breaking changes of (currenlty completely hypothetical) bird4 are.

SuperSandro2000 replied:
> Letting the Package depend on stateVersion is a horrible idea since all changes are completely config related and mkRenamedOptionModule won't tell users that changing this will implicitly also update to the newest major version of bird.

This refers to what? `stateVersion`? Anyone changing stateVersion in their config better knows what they are doing and *is* completely on its own. It is not a supported scenario and completely unsupported.

> If we were to just pin the package there still would be changes:
>
> The name of the systemd service
> The user/group as which it's executed
> the location of the config file
> For most Users this probably isn't relevant, but changing those still could break something somewhere.
> One could now attempt to pin all of those values to the stateVersion but that appears to me like enormous complexicty and a testing and validation nightmare going forward.

That's a fair and the first valid point in this thread. Especially the systemd service name and groups easily get into conflict with secrets.

snaakey commented:
> Edit: I've extended the info text of `lib.mkRemovedOptionModule` to also include that there is a switch to bird3.

I think this should be enough.

herbetom committed:
bird2: fetch src url via https

herbetom commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 366190`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.bird</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>bird2</li>
    <li>bird3</li>
  </ul>
</details>

herbetom commented:
> @snaakey Having a check requiring Users to set the package option when checkConfig is set to false sounds like an intersting idea. I'll look into it.

I'll thought about it again and i think for now we don't need such a check. It's added complexity and during debugging it is sometimes quite helpful to disable it shortly. But since i don't expect the "bird3" default to change anytime soon again it seems unncesary to enforce setting it now already.
But i think if there ever is another major change (bird4 or whatever) it's worth considering to introducing something along those lines then.

herbetom commented:
@SuperSandro2000 Posting the same reply twice is a really strange move. Was that intentional? I replied to you here: https://github.com/NixOS/nixpkgs/pull/366190#discussion_r1923857137

MarcelCoding commented on nixos/modules/services/networking/bird.nix:
```diff
@@ -68,36 +69,38 @@ in
-      name = "bird2";
+    environment.etc."bird/bird.conf".source = pkgs.writeTextFile {
+      name = "bird";
       text = cfg.config;
       checkPhase = optionalString cfg.checkConfig ''
-        ln -s $out bird2.conf
+        ln -s $out bird.conf
         ${cfg.preCheckConfig}
-        ${pkgs.buildPackages.bird}/bin/bird -d -p -c bird2.conf
+        ${pkgs.buildPackages.bird}/bin/bird -d -p -c bird.conf
       '';
     };
```
I think the package option should be used here. I don't know hot this should be handled in regard to cross compilation.

I've tested the behavior in a repl and this should work.
```nix
    environment.etc."bird/bird.conf".source = pkgs.writeTextFile {
      name = "bird";
      text = cfg.config;
      derivationArgs.nativeBuildInputs = [ cfg.package ];
      checkPhase = optionalString cfg.checkConfig ''
        ln -s $out bird.conf
        ${cfg.preCheckConfig}
        bird -d -p -c bird.conf
      '';
    };
```

herbetom replied:
Good point, changed that.

SuperSandro2000 commented on pkgs/top-level/aliases.nix:
```diff
@@ -152,7 +152,7 @@ mapAliases {
   bitcoin-unlimited = throw "bitcoin-unlimited has been removed as it was broken and unmaintained"; # Added 2024-07-15
   bitcoind-unlimited = throw "bitcoind-unlimited has been removed as it was broken and unmaintained"; # Added 2024-07-15
-  bird = bird2; # Added 2025-01-11
+  bird = bird3; # Added 2025-01-11
```
We should keep that as bird2 as we otherwise silently upgrade people that are right now using the alias.

SuperSandro2000 replied:
```suggestion
  bird = bird2; # Added 2025-01-11
```

herbetom replied:
I don't aggree with this last minute change.

IMHO a general "bird" alias should point to the latest release. Not to some random old version where we're going to have to figure out what to do once we drop that version.

Yes, someone might be using `pkgs.bird` in some context and find themself with bird3, but i don't think that would be too bad. IMHO chances are good it just works.
Using `pkgs.bird2` was already an option previously, even if not well documented.

And with the other changes already happening i really would like to upgrade the "bird" refrence now and not down the line once we drop bird2.

Changing/Appending/Adding to the Release Notes might be something worth disccussing to make it more clear that the change is coming but it's even to some degree already included:

> services.bird2 has been renamed to services.bird and **the default bird package has been switched to bird3**. bird2 can still be choosen via the services.bird.package option.

Another option would of course be to just drop the `bird` alias once bird2 is removed. But i think there should be an versionless alias to the latest version like i think it's the cases with other software?

SuperSandro2000 replied:
> services.bird2 has been renamed to services.bird and **the default bird package has been switched to bird3**.

I think this can be easily read either way, like that the package in the module was only changed.

SuperSandro2000 committed:
bird: change alias back to bird2

as we otherwise silently update people using the alias which we do not want in this case

SuperSandro2000 commented:
> @SuperSandro2000 Posting the same reply twice is a really strange move. Was that intentional? I replied to you here: [#366190 (comment)](https://github.com/NixOS/nixpkgs/pull/366190#discussion_r1923857137)

GitHub probably double posted when networking was having a hickup. Happens from time to time.

MarcelCoding commented:
Also with 3.0.1 or still with 3.0.0?

herbetom commented:
@yuyuyureka oh, that's not good. Could you please send a report to the bird Mailing list? https://bird.network.cz/?m_list That way we hopefully get this resolved soon.

But i would like to stay with bird3 as the default as it seems like a more sensible default going forward and due to our use of `mkRemovedOptionModule` it won't suprise anyone.

herbetom commented:
Hi @yuyuyureka,

just wanted to follow up. You've seen that there was a response on the Bird Mailing List with some questions regarding your Setup? Or did you already respond in private or haven't gotten the time yet or whatever?

https://bird.network.cz/pipermail/bird-users/2025-February/018067.html

I just want to make sure this isn't forgotten about.

herbetom commented:
Thanks, good to hear! :)
